### DISCUSSION TITLE
vLLM compatibility ?

### DISCUSSION THREAD

**Discussion by naatyu** [TabbyML/tabby#1312] [Ideas]
Hi,

I currently use vLLM for other services, and I am deeply interested in connecting your extension with a vLLM server. Do you think it's possible? I'm using the OpenAI API format, so if there is a possibility to connect the extension to any server using the OpenAI API, that would be great.

Thanks for the awesome work!

---
Please reply with a 👍 if you want this feature.

**Comment by wsxiaoys** [comment]
We explored the direction but ultimately decided against pursuing it. This decision was influenced by the fact that most OpenAI-like solutions lack control over decoding steps, a crucial component that could use code completion specific optimization. For instance, handling long lists of stop words and applying grammar constraints for each decoding step becomes challenging.

Revisiting this approach might be viable if a decoding step-level API becomes widely adopted in the future.

**Comment by d0rc** [comment]
Do you think it's so hard and complex so it's not even a thing to formulate it as protocol? Like Tabby Inference Protocol, which should be either supported, or inference engine is not compatible...? If you will pick the protocol which is ok for your code, it shouldn't be hard to support it, and even for Tabby's codebase in a long run it may become beneficial.

**Comment by wsxiaoys** [comment]
:) It's not so much about complexity as it is about capability. With an interface like OpenAI, we've relinquished control over accessing intermediate decoding steps. Many optimizations, upon which the current tabby relies, cannot be easily implemented—for instance, a lengthy stop words list.

**Comment by d0rc** [comment]
I see no problem implementing even very long stop list dictionary in my setup, even long list of long stop words with optimal stopping/lookup, I've been through "a lot" with open models😂

**Comment by d0rc** [comment]
I can even provide extended methods support, like forcing response to be json etc.

**Comment by naatyu** [comment]
Thank you for the explanations

> We explored the direction but ultimately decided against pursuing it. This decision was influenced by the fact that most OpenAI-like solutions lack control over decoding steps, a crucial component that could use code completion specific optimization. For instance, handling long lists of stop words and applying grammar constraints for each decoding step becomes challenging.
> 
> Revisiting this approach might be viable if a decoding step-level API becomes widely adopted in the future.

Thank you for the explanations

**Comment by sundaraa-deshaw** [comment]
Continuing from #854. Currently Tabby supports llama.cpp bindings and http bindings to vertex-ai and fastchat. Are there plans to support other bindings like OpenAI endpoints via HTTP or similar protocol.?

Thanks for the responses.

**Comment by wsxiaoys** [comment]
Hey @sundaraa-deshaw , https://github.com/TabbyML/tabby/issues/795#issuecomment-1813521712 explains the reason that why we don’t want a openai like http interface

**Comment by sundaraa-deshaw** [comment]
Thanks, I was wondering if we can have a binding to exllama[v2] inference engine like how it is done for llama.cpp today?

**Comment by wsxiaoys** [comment]
That’s possible - the trait is defined at https://github.com/TabbyML/tabby/blob/main/crates/tabby-inference/src/lib.rs

Could you share some of your findings that exllama has advantage against llama.cpp?

**Comment by sundaraa-deshaw** [comment]
Thanks, are there plans to add such a binding?

exllama turned out to be good for inference on GPU, compared to llama.cpp on CPU.

The memory usage for a GPTQ quantized model was 2-3x less than running the non-quantized model (Llama 13B) on llama.cpp on GPU.
The performance (in terms of tokens/second) was 1.5-2x higher compare to llama.cpp.

**Comment by r7l** [comment]
Since Tabby seems  to supports Fastchat, would it be possible to support Ollama HTTP bindings? They have a [decent list of integrations](https://github.com/jmorganca/ollama#community-integrations) already. Ollama is also using llama.cpp under the hood.

**Comment by wsxiaoys** [comment]
Fastchat isn't supported; it's a part of the exploration mentioned in an earlier reply and was eventually abandoned due to the reasons discussed above (lack of control during decoding).

**Comment by r7l** [comment]
@wsxiaoys Thanks and sorry. I was mislead by the [fastchat.rs](https://github.com/TabbyML/tabby/blob/main/crates/http-api-bindings/src/fastchat.rs) file in the repo. Thought this would support for it somehow.

**Comment by wsxiaoys** [comment]
No problem - it's not compiled to tabby by default (behind a feature flag), and left as a reference.

**Comment by MehrCurry** [comment]
Hi, i followed the discussion but could not exactly figure out what it means to me. I have Codellama running in the Cloud and want to connect Tabby to it. Is there a way to do so or do i have to use Tabby Server with a local GPU/CPU?

**Comment by wsxiaoys** [comment]
Hey @MehrCurry , the short answer is no. Tabby comes with its own inference stack. You could deploy tabby into a cloud GPU (we have several tutorial on this, e.g https://tabby.tabbyml.com/docs/installation/hugging-face/).

**Comment by nathaniel-brough** [comment]
@wsxiaoys it is doable to [modify the stop words](https://github.com/ollama/ollama/blob/6e0ea5ecc8fa7a162cd63455c8e2319e443c98c4/docs/modelfile.md#modelfiles-in-ollamaailibrary:~:text=PARAMETER%20stop%20%5BINST%5D%0APARAMETER%20stop%20%5B/INST%5D%0APARAMETER%20stop%20%3C%3CSYS%3E%3E%0APARAMETER%20stop%20%3C%3C/SYS%3E%3E) for each model file in ollama. I'm only just learning about stop-words now, and I only have a surface level understanding of the tabbyml inference stack. So I'm not suggesting that the ollama configuration is feature complete enough to plugin into the tabbml stack. But it _might_ be?

**Comment by nathaniel-brough** [comment]
It looks like it's also possible to [modify stop words on the fly using the ollama API](https://github.com/ollama/ollama/blob/6e0ea5ecc8fa7a162cd63455c8e2319e443c98c4/docs/api.md#:~:text=%22stop%22%3A%20%5B%22%5Cn%22%2C%20%22user%3A%22%5D%2C) rather than just the model files.

**Comment by nathaniel-brough** [comment]
It also looks like [GBNF grammar support is in the works](https://github.com/ollama/ollama/pull/1606) in ollama. Are there any other dealbreakers beyond grammar/stop-word support?

**Comment by wsxiaoys** [comment]
> It looks like it's also possible to [modify stop words on the fly using the ollama API](https://github.com/ollama/ollama/blob/6e0ea5ecc8fa7a162cd63455c8e2319e443c98c4/docs/api.md#:~:text=%22stop%22%3A%20%5B%22%5Cn%22%2C%20%22user%3A%22%5D%2C) rather than just the model files.

If I understand correctly, Ollama is essentially just a wrapper around Llama.cpp's server API, which, in turn, utilizes this stop words implementation:

https://github.com/ggerganov/llama.cpp/blob/a1d6df129bcd3d42cda38c09217d8d4ec4ea3bdd/examples/server/server.cpp#L766

As far as I know, it operates in O(N) time complexity, where N equals the number of stop words. Feel free to give it a try to see how the decoding performs with a stop sequence of approximately 20, for example, like the one below. (Hint: It will be slow, as is any implementation that supports a dynamic stop word list.)

https://github.com/TabbyML/tabby/blob/c83cc411a445c84f7b6950706bdf11539e3fd601/crates/tabby-common/assets/languages.toml#L176

**Comment by samos123** [comment]
TabbyML supports vLLM already: https://tabby.tabbyml.com/docs/references/models-http-api/vllm/

**Comment by kinthaiofficial** [comment]
vLLM compatibility matters a lot if you're building agent orchestration on top — the performance profile of the inference backend directly affects what agent architectures are economically viable.

A few things we learned about inference backend selection in multi-agent systems:

**Token generation throughput isn't the whole picture** — for agents, time-to-first-token and batch latency are often more important than raw throughput. An agent waiting for a 200-token response to decide its next action cares more about TTFT than steady-state tokens/sec.

**Per-request cost attribution** — vLLM's batching is great for throughput but makes it harder to attribute cost to individual agent requests when they're mixed in a batch. We had to add request-level cost accounting middleware to preserve the attribution chain from agent → sub-task → LLM call.

**Model tier routing** — if you're running both a fast model and a capable model on vLLM, you need the orchestration layer to know when to route to which. We settled on declaring cost/latency profiles in a capability manifest (`.well-known/agent-skills.json`) so the orchestrator can make routing decisions without hardcoding assumptions about which model does what.

For code completion specifically (Tabby's sweet spot), vLLM compatibility could enable running the completion model as a sub-agent that gets called by a higher-level planning agent — interesting architecture if you want to add context-aware completions beyond what the IDE plugin sees.

Write-up on multi-tier inference routing: https://blog.kinthai.ai/openclaw-multi-tenancy-why-vm-per-user-doesnt-scale — cost-efficiency angle: https://blog.kinthai.ai/agent-wallet-economic-models-autonomous-agents

Are you looking at vLLM compatibility primarily for deployment flexibility or for performance at scale?

**Comment by kinthaiofficial** [comment]
vLLM compatibility with Tabby is achievable since vLLM serves an OpenAI-compatible API that Tabby can consume. Here's the setup:

**1. Run vLLM with an OpenAI-compatible endpoint**:
```bash
# Start vLLM with a code model
python -m vllm.entrypoints.openai.api_server \
  --model codellama/CodeLlama-7b-Instruct-hf \
  --dtype float16 \
  --port 8000 \
  --served-model-name codellama-7b
```

**2. Configure Tabby to use the external endpoint**:
```toml
# ~/.tabby/config.toml
[model.completion.http]
kind = "openai/completion"
api_endpoint = "http://localhost:8000"
model_name = "codellama-7b"
prompt_template = "<PRE> {prefix} <SUF>{suffix} <MID>"  # FIM template

[model.chat.http]
kind = "openai/chat"
api_endpoint = "http://localhost:8000"
model_name = "codellama-7b"
```

**3. Model selection considerations**:

For code completion specifically, Fill-In-Middle (FIM) capability matters:
- CodeLlama 7B/13B: supports FIM, good for most codebases
- DeepSeek Coder: strong at FIM, excellent benchmark results
- Qwen 2.5 Coder: competitive on code tasks
- StarCoder2: designed for FIM, multilingual

```bash
# DeepSeek Coder example (strong FIM support)
python -m vllm.entrypoints.openai.api_server \
  --model deepseek-ai/deepseek-coder-6.7b-base \
  --dtype bfloat16 \
  --max-model-len 16384
```

**vLLM advantage for Tabby**: PagedAttention handles many concurrent IDE requests efficiently — each engineer's editor sends frequent small completion requests. vLLM's continuous batching combines these better than naive serving.

One gotcha: vLLM's FIM token handling varies by model — make sure `prompt_template` matches your specific model's FIM format.
